File: backend/app/schemas/coding.py
"""
Pydantic schemas for the Coding Interview Platform API.
"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Question schemas
# ---------------------------------------------------------------------------

class CodingQuestionListItem(BaseModel):
    id: int
    title: str
    slug: str
    difficulty: str
    category: str
    tags: list[str]
    optimal_time_complexity: str | None = None
    optimal_space_complexity: str | None = None

    class Config:
        from_attributes = True


class CodingQuestionDetail(CodingQuestionListItem):
    description: str
    constraints: list[str]
    examples: list[dict[str, Any]]
    default_code: dict[str, str]
    test_cases_public: list[dict[str, Any]]
    hints: list[str]

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Session schemas
# ---------------------------------------------------------------------------

class CodingSessionCreate(BaseModel):
    question_id: int
    language: str = Field("python", pattern=r"^(python|javascript|java|cpp|c)$")


class CodingSessionResponse(BaseModel):
    id: int
    question_id: int
    language: str
    status: str
    code_snapshot: str | None = None
    timer_seconds_elapsed: int
    started_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class CodingSessionDetail(CodingSessionResponse):
    question: CodingQuestionDetail

    class Config:
        from_attributes = True


class CodeSnapshotSave(BaseModel):
    code: str
    timer_seconds_elapsed: int
    language: str = Field("python", pattern=r"^(python|javascript|java|cpp|c)$")


# ---------------------------------------------------------------------------
# Run / Submit schemas
# ---------------------------------------------------------------------------

class CodeRunRequest(BaseModel):
    code: str
    language: str = Field("python", pattern=r"^(python|javascript|java|cpp|c)$")
    custom_input: str | None = None  # overrides default test cases when present


class TestCaseResult(BaseModel):
    test_case_index: int
    input: str
    expected_output: str
    actual_output: str
    passed: bool
    runtime_ms: float | None = None


class CodeRunResponse(BaseModel):
    verdict: str
    passed_tests: int
    total_tests: int
    runtime_ms: float | None = None
    memory_kb: float | None = None
    stdout: str | None = None
    stderr: str | None = None
    test_results: list[TestCaseResult]
    submission_id: int


class CodeSubmitRequest(BaseModel):
    code: str
    language: str = Field("python", pattern=r"^(python|javascript|java|cpp|c)$")


class CodeSubmitResponse(CodeRunResponse):
    """Same as run but always runs against all hidden test cases."""
    pass


# ---------------------------------------------------------------------------
# Submission history
# ---------------------------------------------------------------------------

class CodeSubmissionSummary(BaseModel):
    id: int
    language: str
    is_run: bool
    verdict: str | None = None
    passed_tests: int
    total_tests: int
    runtime_ms: float | None = None
    memory_kb: float | None = None
    submitted_at: datetime

    class Config:
        from_attributes = True


class CodeSubmissionDetail(CodeSubmissionSummary):
    code: str
    execution: "CodeExecutionResponse | None" = None
    feedback: "CodingFeedbackResponse | None" = None

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Execution result
# ---------------------------------------------------------------------------

class CodeExecutionResponse(BaseModel):
    id: int
    stdout: str | None = None
    stderr: str | None = None
    execution_time_ms: float | None = None
    memory_kb: float | None = None
    exit_code: int
    test_results: list[dict[str, Any]]
    executed_at: datetime

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# AI Review
# ---------------------------------------------------------------------------

class CodingFeedbackResponse(BaseModel):
    id: int
    submission_id: int
    quality_score: float
    time_complexity: str
    space_complexity: str
    bugs: list[dict[str, Any]]
    suggestions: list[str]
    best_practices: list[str]
    interview_notes: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


# Required for forward references in CodeSubmissionDetail
CodeSubmissionDetail.model_rebuild()

File: frontend/src/types/coding.ts
// TypeScript interfaces for the Coding Interview Platform

export type Difficulty = "easy" | "medium" | "hard";
export type Language = "python" | "javascript" | "java" | "cpp" | "c";
export type Verdict =
  | "accepted"
  | "wrong_answer"
  | "runtime_error"
  | "time_limit_exceeded"
  | "compilation_error";
export type SessionStatus = "in_progress" | "completed" | "abandoned";

// -------------------------------------------------------------------
// Questions
// -------------------------------------------------------------------

export interface CodingQuestionListItem {
  id: number;
  title: string;
  slug: string;
  difficulty: Difficulty;
  category: string;
  tags: string[];
  optimal_time_complexity: string | null;
  optimal_space_complexity: string | null;
}

export interface CodingQuestionExample {
  input: string;
  output: string;
  explanation?: string;
}

export interface CodingQuestionDetail extends CodingQuestionListItem {
  description: string;
  constraints: string[];
  examples: CodingQuestionExample[];
  default_code: Record<Language, string>;
  test_cases_public: Array<{ input: string; output: string }>;
  hints: string[];
}

// -------------------------------------------------------------------
// Sessions
// -------------------------------------------------------------------

export interface CodingSessionResponse {
  id: number;
  question_id: number;
  language: Language;
  status: SessionStatus;
  code_snapshot: string | null;
  timer_seconds_elapsed: number;
  started_at: string;
  updated_at: string;
}

export interface CodingSessionDetail extends CodingSessionResponse {
  question: CodingQuestionDetail;
}

export interface CodeSnapshotSave {
  code: string;
  language: Language;
  timer_seconds_elapsed: number;
}

// -------------------------------------------------------------------
// Run / Submit
// -------------------------------------------------------------------

export interface TestCaseResult {
  test_case_index: number;
  input: string;
  expected_output: string;
  actual_output: string;
  passed: boolean;
  runtime_ms: number | null;
}

export interface CodeRunResponse {
  verdict: Verdict;
  passed_tests: number;
  total_tests: number;
  runtime_ms: number | null;
  memory_kb: number | null;
  stdout: string | null;
  stderr: string | null;
  test_results: TestCaseResult[];
  submission_id: number;
}

export type CodeSubmitResponse = CodeRunResponse;

// -------------------------------------------------------------------
// Submission history
// -------------------------------------------------------------------

export interface CodeSubmissionSummary {
  id: number;
  language: Language;
  is_run: boolean;
  verdict: Verdict | null;
  passed_tests: number;
  total_tests: number;
  runtime_ms: number | null;
  memory_kb: number | null;
  submitted_at: string;
}

export interface CodeExecutionResult {
  id: number;
  stdout: string | null;
  stderr: string | null;
  execution_time_ms: number | null;
  memory_kb: number | null;
  exit_code: number;
  test_results: TestCaseResult[];
  executed_at: string;
}

export interface CodingFeedback {
  id: number;
  submission_id: number;
  quality_score: number;
  time_complexity: string;
  space_complexity: string;
  bugs: Array<{ severity: "critical" | "major" | "minor"; description: string; line_hint?: string }>;
  suggestions: string[];
  best_practices: string[];
  interview_notes: string | null;
  created_at: string;
}

export interface CodeSubmissionDetail extends CodeSubmissionSummary {
  code: string;
  execution: CodeExecutionResult | null;
  feedback: CodingFeedback | null;
}
